hrms/hr/doctype/appraisal/appraisal.py

```diff
@@ -178,9 +179,35 @@ def calculate_avg_feedback_score(self, update=False):
 			self.db_update()
 
 	def calculate_final_score(self):
-		final_score = (flt(self.total_score) + flt(self.avg_feedback_score) + flt(self.self_score)) / 3
+		if self.appraisal_cycle:
+			apraisal_cycle_doc = frappe.get_doc("Appraisal Cycle", self.appraisal_cycle)
+			employee_doc = frappe.get_doc("Employee", self.employee)
 
-		self.final_score = flt(final_score, self.precision("final_score"))
+			formula = apraisal_cycle_doc.final_score_formula
+			based_on_formula = apraisal_cycle_doc.calculate_final_score_based_on_formula
+
+			appraisal_cycle_fields = apraisal_cycle_doc.as_dict()
+			employee_fields = employee_doc.as_dict()
+
+			sanitized_formula = sanitize_expression(formula)
+
+			vars = {
+				"goal_score": flt(self.total_score),
+				"average_feedback_score": flt(self.avg_feedback_score),
+				"self_appraisal_score": flt(self.self_score),
+			}
+
+			vars.update(appraisal_cycle_fields)
+			vars.update(employee_fields)
+
+			vars = {key: flt(value) for (key, value) in vars.items()}
```

Comment:
```suggestion
			sanitized_formula = sanitize_expression(formula)

			data = frappe._dict({
				"goal_score": flt(self.total_score),
				"average_feedback_score": flt(self.avg_feedback_score),
				"self_appraisal_score": flt(self.self_score),
			})

			data.update(appraisal_cycle_doc.as_dict())
			data.update(self.as_dict())
			data.update(employee_doc.as_dict())
```

- updating as_dict() directly here saving some extra variables
- lets call vars -> data? 
- Appraisal fields should also be available for eval -> added self.as_dict()
- Do we really need the flt() casting on line number 203. Not all fields in vars are float right?

Reply:
added flt because, adding some field that wasnt a float in expression was throwing incompatible data types error